# Wstęp do Sztucznej Inteligencji - rok akademicki 2021/2022

Przed rozpoczęciem pracy z notatnikiem zmień jego nazwę zgodnie z wzorem: `NrAlbumu_Nazwisko_Imie_PoprzedniaNazwa`.

Przed wysłaniem notatnika upewnij się, że rozwiązałeś wszystkie zadania/ćwiczenia.

# Temat: Wnioskowanie oparte na wiedzy niepewnej. Wnioskowanie bayesowskie
Zapoznaj się z treścią niniejszego notatnika czytając i wykonując go komórka po komórce. Wykonaj napotkane zadania/ćwiczenia.

In [ ]:
import pymc3 as pm
import arviz as az
import numpy as np
print('Pymc3 version:', pm.__version__)
print('Arviz version:', az.__version__)
print('Numpy version:', np.__version__)

Pymc3 version: 3.11.4
Arviz version: 0.11.4
Numpy version: 1.21.5


## Zadanie 1  (obowiązkowe, 5pkt.)
Zapożyczone z ćwiczeń do wykładu [Arabas, Cichosz](http://wazniak.mimuw.edu.pl/index.php?title=Sztuczna_inteligencja/SI_%C4%86wiczenia_4)

### Zamodeluj i odpowiedz na pytania.

W śledztwie dotyczącym zabójstwa inspektor Bayes rozważa dwie hipotezy:

- $H_1$ główny podejrzany zabił,
- $H_2$ główny podejrzany nie zabił, 

oraz następujące możliwe fakty:

- $E_1$ na miejscu zbrodni znaleziono odciski palców głównego podejrzanego,
- $E_2$ główny podejrzany nie ma alibi na czas popełnienia zabójstwa,
- $E_3$ główny podejrzany miał motyw zabicia ofiary,
- $E_4$ główny podejrzany był widziany w sądziedztwie miejsca, w którym mieszka nielegalny handlarz bronią,
- $E_5$ świadek zbrodni podał rysopis zabójcy nie pasujący do głównego podejrzanego. 

Zależności między takimi faktami a hipotezami opisują następujące prawdopodobieństwa:

$P(E_1|H_1)=0.7,\qquad P(E_1|H_2)=0.3,$

$P(E_2|H_1)=0.8,\qquad P(E_2|H_2)=0.4,$

$P(E_3|H_1)=0.9,\qquad P(E_3|H_2)=0.5,$

$P(E_4|H_1)=0.4,\qquad P(E_4|H_2)=0.2,$

$P(E_5|H_1)=0.2,\qquad P(E_5|H_2)=0.4.$ 

__W którym przypadku prawdopodobieństwo popełnienia zabójstwa byłoby największe?__

1. Gdyby znaleziono na miejscu zbrodni jego odciski palców.
2. Gdyby stwierdzono, że nie miał alibi i miał motyw.
3. Gdyby znaleziono na miejscu zbrodni jego odciski palców oraz stwierdzono, że był widziany w sąsiedztwie miejsca, w którym mieszka nielegalny handlarz bronią, ale świadek zbrodni podał rysopis zabójcy nie pasujący do głównego podejrzanego.

### TWÓJ PROGRAM:

In [4]:
#YOUR CODE HERE
import pymc3 as pm
import arviz as az
import numpy as np

with pm.Model() as morderstwo_model:
    zabil_H1 = pm.Bernoulli('Podejrzany zabił', 0.5)
    
    p_E1 = pm.Deterministic('p_E1', pm.math.switch(zabil_H1, 0.7, 0.3))
    e1 = pm.Bernoulli('E1', p_E1)
    
    p_E2 = pm.Deterministic('p_E2', pm.math.switch(zabil_H1, 0.8, 0.4))
    e2 = pm.Bernoulli('E2', p_E2)
    
    p_E3 = pm.Deterministic('p_E3', pm.math.switch(zabil_H1, 0.9, 0.5))
    e3 = pm.Bernoulli('E3', p_E3)
    
    p_E4 = pm.Deterministic('p_E4', pm.math.switch(zabil_H1, 0.4, 0.2))
    e4 = pm.Bernoulli('E4', p_E4)
    
    p_E5 = pm.Deterministic('p_E5', pm.math.switch(zabil_H1, 0.2, 0.4))
    e5 = pm.Bernoulli('E5', p_E5)
    
    trace = pm.sample(200000, chains=1)
    
    p_przypadek_1 = (trace['E1'] * trace['Podejrzany zabił']).sum() / (trace['E1'] * trace['Podejrzany zabił'] + trace['E1'] * np.logical_not(trace['Podejrzany zabił'])).sum()
    print('Przypadek 1:', p_przypadek_1)
    
    p_przypadek_2 = (trace['E2'] * trace['E3'] * trace['Podejrzany zabił']).sum() / (trace['E2'] * trace['E3'] * trace['Podejrzany zabił'] + trace['E2'] * trace['E3'] * np.logical_not(trace['Podejrzany zabił'])).sum()
    print('Przypadek 2:', p_przypadek_2)
    
    p_przypadek_3 = (trace['E1'] * trace['E4'] * trace['E5'] * trace['Podejrzany zabił']).sum() / (trace['E1'] * trace['E4'] * trace['E5'] * trace['Podejrzany zabił'] + trace['E1'] * trace['E4'] * trace['E5'] * np.logical_not(trace['Podejrzany zabił'])).sum()
    print('Przypadek 3:', p_przypadek_3)
    
    print(trace['E2'].sum())

/usr/local/lib/python3.7/dist-packages/ipykernel_launcher.py:24: FutureWarning: In v4.0, pm.sample will return an `arviz.InferenceData` object instead of a `MultiTrace` by default. You can pass return_inferencedata=True or return_inferencedata=False to be safe and silence this warning.
Sequential sampling (1 chains in 1 job)
BinaryGibbsMetropolis: [Podejrzany zabił, E1, E2, E3, E4, E5]


Sampling 1 chain for 1_000 tune and 200_000 draw iterations (1_000 + 200_000 draws total) took 131 seconds.
Only one chain was sampled, this makes it impossible to run some convergence checks


Przypadek 1: 0.7035549147317656
Przypadek 2: 0.7847957517223478
Przypadek 3: 0.6998622071902794
120294


### ODPOWIEDŹ:

YOUR ANSWER HERE

Przypadek 1: 0.7035549147317656
Przypadek 2: 0.7847957517223478
Przypadek 3: 0.6998622071902794

Największe prawdopodobieństwo zabójstwa byłoby w przypdaku nr 2, czyli gdyby stwierdzono, że nie miał alibi i miał motyw.

## Zadanie 2  (obowiązkowe, 5pkt.)

### Zamodeluj i odpowiedz na pytania.
System alarmowy w mieszkaniu, reaguje na włamania oraz, niestety, również na drobne trzęsienia (ziemi). Sąsiedzi John i Mary są umówieni, żeby zadzwonić do właściciela gdy usłyszą alarm. John jest nadgorliwy i bierze różne zdarzenia (np. dzwonek telefonu) za sygnał alarmowy (i wtedy zawsze dzwoni). Mary rozpoznaje alarm poprawnie, lecz często słucha głośnej muzyki i może go w ogóle nie usłyszeć. 

Sieć przekonań dla systemu alarmowego wygląda następująco:
![bsiec.PNG](http://torus.uck.pk.edu.pl/~amarsz/images/bsiec.PNG)

__Jakie jest prawdopodobieństwo, że:__
1. włączy się alarm?
2. doszło do włamanie jeśli wiadom, że włączył się alarm?
3. zdarzyło się trzęsienie ziemi jeśli wiadomo, żę włączył się alarm?
1. w razie włamania ktoś zadzwoni?
2. zawiadomienie o włamaniu jest fałszywe?
3. rozległ się alarm, przy czym nie wystąpiło ani trzęsienie ziemi ani włamanie, ale oboje John i Mary zadzwonili? (prawd. bezwarunkowe)

TWÓJ PROGRAM:

In [3]:
#YOUR CODE HERE

import pymc3 as pm
import numpy as np

with pm.Model() as alarm:
    
    earthquake = pm.Bernoulli('ear', 0.002)
    burglary = pm.Bernoulli('bur', 0.001)
    
    p_alarm = pm.Deterministic('p_alarm', pm.math.switch(burglary, pm.math.switch(earthquake, 0.95, 0.94), pm.math.switch(earthquake, 0.29, 0.001)))
    alarm = pm.Bernoulli('alarm', p_alarm)
    
    p_John = pm.Deterministic('p_John', pm.math.switch(alarm, 0.9, 0.05))
    john = pm.Bernoulli('john', p_John)
    p_Mary = pm.Deterministic('p_Mary', pm.math.switch(alarm, 0.7, 0.01))
    mary = pm.Bernoulli('mary', p_Mary)
    
    ilosc_symulacji = 200000
    trace = pm.sample(ilosc_symulacji, chains=1)
    
    zad_1 = trace['alarm'].sum() / ilosc_symulacji
    print('Pytanie 1: ', zad_1)
    
    zad_2 = (trace['alarm'] * trace['bur']).sum() / trace['alarm'].sum()
    print('Pytanie 2: ', zad_2) 
    
    zad_3 = (trace['alarm'] * trace['ear']).sum() / trace['alarm'].sum()
    print('Pytanie 3: ', zad_3) 
    
    zad_4 = ((trace['john'] + trace['mary'] - trace['john'] * trace['mary']) * trace['bur']).sum() / trace['bur'].sum()
    print('Pytanie 4: ', zad_4)
    
    zad_5 = ((np.logical_or(trace['mary'], trace['john'])) * np.logical_not(trace['bur'])).sum() / np.logical_not(trace['bur']).sum()
    print('Pytanie 5: ', zad_5)
    
    zad_6 = (trace['alarm'] * trace['john'] * trace['mary'] * np.logical_not(trace['bur']) * np.logical_not(trace['ear'])).sum() / ilosc_symulacji
    print('Pytanie 6: ', zad_6)
    

/usr/local/lib/python3.7/dist-packages/ipykernel_launcher.py:20: FutureWarning: In v4.0, pm.sample will return an `arviz.InferenceData` object instead of a `MultiTrace` by default. You can pass return_inferencedata=True or return_inferencedata=False to be safe and silence this warning.
Sequential sampling (1 chains in 1 job)
BinaryGibbsMetropolis: [ear, bur, alarm, john, mary]


Sampling 1 chain for 1_000 tune and 200_000 draw iterations (1_000 + 200_000 draws total) took 114 seconds.
Only one chain was sampled, this makes it impossible to run some convergence checks


Pytanie 1:  0.002385
Pytanie 2:  0.3668763102725367
Pytanie 3:  0.24737945492662475
Pytanie 4:  0.9130434782608695
Pytanie 5:  0.06048064219081555
Pytanie 6:  0.00059


### ODPOWIEDŹ:

YOUR ANSWER HERE

Pytanie 1:  0.002385

Pytanie 2:  0.3668763102725367

Pytanie 3:  0.24737945492662475

Pytanie 4:  0.9130434782608695

Pytanie 5:  0.06048064219081555

Pytanie 6:  0.00059
1. Prawdopodobieństwo, że włączy się alarm: 0.002385


2.  Prawdopodobieństwo, że doszło do włamanie jeśli wiadom, że włączył się alarm: 0.3668763102725367

3.  Prawdopodobieństwo, że zdarzyło się trzęsienie ziemi jeśli wiadomo, żę włączył się alarm: 0.24737945492662475

4.  Prawdopodobieństwo, że w razie włamania ktoś zadzwoni: 0.9130434782608695

5.  Prawdopodobieństwo, że zawiadomienie o włamaniu jest fałszywe: 0.06048064219081555

6.  Prawdopodobieństwo, że rozległ się alarm, przy czym nie wystąpiło ani trzęsienie ziemi ani włamanie, ale oboje John i Mary zadzwonili? (prawd. bezwarunkowe: 0.00059

__UWAGA:__ Zwróć uwagę na wielkości podanych prawdopodobieńst aby dobarć odpowiednią liczbę symulacji.

&copy; Katedra Informatyki, Politechnika Krakowska